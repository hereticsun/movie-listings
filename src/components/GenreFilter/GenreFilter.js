import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { selectGenres } from '../../actions/genres';

import styles from './GenreFilter.module.css';

export class GenreFilter extends Component {
  render() {
    const {genres, movies} = this.props;

    let movieGenres = [];
    let availableGenres=[];
    if(movies && movies.moviesList) {
      movies.moviesList.map(movie => {
        return movieGenres = movieGenres.concat(movie.genre_ids)
      });
      movieGenres = [...new Set(movieGenres)];

      if(Boolean(genres) && genres.length > 0) {
        availableGenres = genres.filter(genre => {
          return movieGenres.find(id => id === genre.id);
        });
      }
    }

    const handleFilter = () => {
      // Create an array of selected genre IDs
      if(Boolean(genres) && genres.length > 0) {
        const selectedGenres = Array.from(document.querySelectorAll('.genreFilterInput:checked'));
        let selectedGenreIDs = [];

        if(selectedGenres) {
          selectedGenres.map(genre => {
            return selectedGenreIDs.push(Number(genre.value));
          });
        }
        this.props.selectGenres(selectedGenreIDs);
      }
    };

    return (
      <section>
        <h2 className={styles.title}>
          Filter by genre
        </h2>
        <form>
          {
            (availableGenres.length > 0) ?
            availableGenres.map(genre => (
                <div key={genre.id}>
                  <label className={styles.label} onClick={handleFilter}>
                    <input
                      className={`genreFilterInput ${styles.input}`}
                      type="checkbox"
                      value={genre.id}
                      
                    />
                    {genre.name}
                  </label>
                </div>
              ))
            : null
          }
        </form>
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
    genres: state.genres.genres,
    movies: state.movies,
  };
}

GenreFilter.propTypes = {
  genres: PropTypes.array,
};

export default connect(mapStateToProps, {selectGenres})(GenreFilter);
