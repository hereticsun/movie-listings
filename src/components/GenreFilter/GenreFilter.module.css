:root {
  color: #fff;
}

.title {
  font-size: 20px;
  font-weight: 600;
}

.label {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  margin-bottom: 8px;
}
.input {
  margin-right: 8px;
}
