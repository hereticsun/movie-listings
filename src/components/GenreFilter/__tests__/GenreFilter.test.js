import React from 'react';
import { shallow } from 'enzyme';
import { GenreFilter } from '../GenreFilter';

describe('GenreFilter', () => {
  const props = {
    genres: [
      {
        id: 1,
        name: 'Action'
      },
      {
        id: 2,
        name: 'Adventure'
      }
    ],
    movies: {
      moviesList: [
        {
          id: 101,
          genre_ids: [1]
        },
        {
          id: 102,
          genre_ids: [2]
        }
      ]
    },
    selectGenres: jest.fn(),
  };

  const element = shallow(<GenreFilter {...props}/>);

  it('displays the list of genres for the listed movies', () => {
    expect(element).toMatchSnapshot();
    expect(element.find('input')).toHaveLength(props.genres.length);
  });
});
