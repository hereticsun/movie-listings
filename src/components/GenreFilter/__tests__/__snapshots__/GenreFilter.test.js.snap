// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreFilter displays the list of genres for the listed movies 1`] = `ShallowWrapper {}`;
